### PR TITLE
Integrate Engine with decoupled Translog interfaces

### PR DESCRIPTION
**Pull request by Bukhtawar**
Signed-off-by: Bukhtawar Khan <bukhtawa@amazon.com>

### Description
The PR aims to Integrate Engine with decoupled Translog interfaces introduced in #3638  . This is the second part of introducing new interfaces and implementation as a part of decoupling. The overall proposal can be reviewed at https://github.com/opensearch-project/OpenSearch/pull/3471
Breaks down the main decouple PR into https://github.com/opensearch-project/OpenSearch/pull/3471 Pluggable translog work

 
### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/3241 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/engine/InternalEngine.java:297]: Contemplating introducing a TranslogFactory

File: server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:106]: I am wondering if `TranslogManager` should implement `Closeable` as well since it manages closeable resource (like `Translog`), wdyt?

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:106]: Thats a good point however there are multiple places where the Engine is consuming the underlying `Translog` directly, which I plan on getting rid of in subsequent PRs as a part of moving Translog to a module(this PR would become too big to review otherwise). 

https://github.com/opensearch-project/OpenSearch/blob/304d830275fdbe5f8095a7285ca66266987c727a/server/src/main/java/org/opensearch/index/translog/TranslogManager.java#L102

Once complete I will make the suggested change

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:106]: Opened https://github.com/opensearch-project/OpenSearch/issues/3709

File: server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:161]: Thanks @Bukhtawar  I went through a few times, I have nothing else to comment on

**Comment by Bukhtawar** [comment]
start gradle check

File: server/src/main/java/org/opensearch/index/translog/NoOpTranslogManager.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/translog/NoOpTranslogManager.java:96]: FYI (new change set), the `NoOpTranslogManager` is effectively unusable since every single caller assumes it will never return `null` for `Translog getTranslog()`. Hopefully, `NoOpTranslogManager` will never be used but you may think about introducing `NoOpTranslog`

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/index/translog/NoOpTranslogManager.java:96]: That's correct the `NoOpTranslog` is already being worked upon as a part of #3242.

**Comment by Bukhtawar** [comment]
start gradle check

File: server/src/main/java/org/opensearch/index/engine/Engine.java
Comment (nknize) [inline server/src/main/java/org/opensearch/index/engine/Engine.java:171]: :tada:  Great to see this decoupling

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java
Comment (nknize) [inline server/src/main/java/org/opensearch/index/engine/InternalEngine.java:338]: Not a big fan of this indirection. I think refactoring utility methods like `getMaxSeqNo` directly to the `TranslogManager` instead of invoking `getTranslog` everywhere is cleaner?

Comment (nknize) [reply server/src/main/java/org/opensearch/index/engine/InternalEngine.java:338]: Update: I just noticed the previous review comments. Looks like you're planning to decouple this further in follow up PRs. So disregard.

File: server/src/test/java/org/opensearch/index/engine/InternalEngineTests.java
Comment (nknize) [inline server/src/test/java/org/opensearch/index/engine/InternalEngineTests.java:764]: This this indirection to translogManager() temporary and planned to be cleaned up in the follow on PRs?

**Comment by Bukhtawar** [comment]
> One general question. Is the point of having TranslogManager as an interface because we want to be able to create other translog managers (e.g., kafka manager)?

Thanks @nknize 
The idea was to free Engine of the translog responsibilities and create `NoOpTranslogManager` to be able to create an `Engine` which doesn't support translog operations. The first part of the PR was https://github.com/opensearch-project/OpenSearch/pull/3638 where we introduced newer interfaces. For remote extensions(kafka) we have #3242 which is being worked upon. The idea there is to decouple `TranslogWriter` from it's underlying store. I will be raising that PR soon.

**Comment by peterzhuamazon** [comment]
Hi @Bukhtawar 
Gradle check fixed. Sorry for confusion.
I rebased your branch and pushed so PR re-run checks now, thanks.

**Comment by nknize** [comment]
>  Is the point of having TranslogManager as an interface because we want to be able to create other translog managers (e.g., kafka manager)?
>> The idea was to free Engine of the translog responsibilities and create various flavours

so, yes? :smile: 

> For remote extensions(kafka) we have #3242 which is being worked upon.

I presume this will create a new `KafkaTranslogManager`? 

I'm trying to glean from this incremental PR if the implementation design is to have 1:1 concrete `TranslogManager` implementations per translog mechanism. So if a user wants to use Kafka as the Translog we, OpenSearch, would provide a `KafkaTranslogManager` for handling the kafka specific logic?

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java
Comment (mch2) [inline server/src/main/java/org/opensearch/index/engine/InternalEngine.java:283]: this is a nit - I'm wondering if we should leave it up to the engine to determine if the error is tragic or not instead of explicitly catching `AlreadyClosedException` in `InternalTranslogManager`?  Then we don't need this separation in `TranslogEventListener`.

**Review by mch2 — APPROVED** [review]
Have only a nit, lgtm.

**Comment by Bukhtawar** [comment]
> I'm trying to glean from this incremental PR if the implementation design is to have 1:1 concrete TranslogManager implementations per translog mechanism. So if a user wants to use Kafka as the Translog we, OpenSearch, would provide a KafkaTranslogManager for handling the kafka specific logic?

The idea is definitely to support streaming stores like kafka and other blob stores to start with. At this point honestly, I am still contemplating if we should introduce a new low level abstraction or continue with TranslogManger as the abstraction itself.

I'll be shortly putting forth a proposal and open up a discussion

**Pull request merged**
